apis/core.oam.dev/v1beta1/resourcetracker_types.go

```diff
@@ -0,0 +1,274 @@
+func (in ManagedResource) ResourceKey() string {
+	gv, kind := in.GroupVersionKind().ToAPIVersionAndKind()
+	return strings.Join([]string{gv, kind, in.Cluster, in.Namespace, in.Name}, "/")
+}
+
+// ComponentKey computes the key for the component which managed resource belongs to
+func (in ManagedResource) ComponentKey() string {
+	return strings.Join([]string{in.Env, in.Component}, "/")
+}
+
+// UnmarshalTo unmarshal ManagedResource into target object
+func (in ManagedResource) UnmarshalTo(obj interface{}) error {
+	if in.Data == nil || in.Data.Raw == nil {
+		return errors.ManagedResourceHasNoDataError{}
+	}
+	return json.Unmarshal(in.Data.Raw, obj)
+}
+
+// ToUnstructured converts managed resource into unstructured
+func (in ManagedResource) ToUnstructured() *unstructured.Unstructured {
+	obj := &unstructured.Unstructured{}
+	obj.SetGroupVersionKind(in.GroupVersionKind())
+	obj.SetName(in.Name)
+	if in.Namespace != "" {
+		obj.SetNamespace(in.Namespace)
+	}
+	oam.SetCluster(obj, in.Cluster)
+	return obj
+}
+
+// ToUnstructuredWithData converts managed resource into unstructured and unmarshal data
+func (in ManagedResource) ToUnstructuredWithData() (*unstructured.Unstructured, error) {
+	obj := in.ToUnstructured()
+	if err := in.UnmarshalTo(obj); err != nil {
+		if errors2.Is(err, errors.ManagedResourceHasNoDataError{}) {
+			return nil, err
+		}
+	}
+	return obj, nil
+}
+
+// ResourceTrackerStatus define the status of resourceTracker
+// For backward-compatibility
+type ResourceTrackerStatus struct {
+	// Deprecated
+	TrackedResources []common.ClusterObjectReference `json:"trackedResources,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ResourceTrackerList contains a list of ResourceTracker
+type ResourceTrackerList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []ResourceTracker `json:"items"`
+}
+
+func (in *ResourceTracker) findMangedResourceIndex(mr ManagedResource) int {
+	for i, _mr := range in.Spec.ManagedResources {
+		if mr.ClusterObjectReference.Equal(_mr.ClusterObjectReference) {
+			return i
+		}
+	}
+	return -1
+}
+
+// AddManagedResource add object to managed resources, if exists, update
+func (in *ResourceTracker) AddManagedResource(rsc client.Object, metaOnly bool) error {
+	gvk := rsc.GetObjectKind().GroupVersionKind()
+	mr := ManagedResource{
+		ClusterObjectReference: common.ClusterObjectReference{
+			ObjectReference: v1.ObjectReference{
+				APIVersion: gvk.GroupVersion().String(),
+				Kind:       gvk.Kind,
+				Name:       rsc.GetName(),
+				Namespace:  rsc.GetNamespace(),
+			},
+			Cluster: oam.GetCluster(rsc),
+		},
+		OAMObjectReference: common.NewOAMObjectReferenceFromObject(rsc),
+		Deleted:            false,
+	}
+	if !metaOnly {
+		bs, err := json.Marshal(rsc)
+		if err != nil {
+			return err
+		}
```

Comment:
does it necessary to marshel it?

Why not just store the object in `&runtime.RawExtension{Object: rsc}`

Reply:
Right, will fix it.